crates/uv/src/commands/project/lock.rs

```diff
@@ -180,59 +179,129 @@ pub(super) async fn do_lock(
     let git = GitResolver::from_refs(git);
 
-    // Create a build dispatch.
-    let build_dispatch = BuildDispatch::new(
-        &client,
-        cache,
-        interpreter,
-        index_locations,
-        &flat_index,
-        &index,
-        &git,
-        &in_flight,
-        index_strategy,
-        setup_py,
-        config_setting,
-        build_isolation,
-        link_mode,
-        build_options,
-        exclude_newer,
-        concurrency,
-        preview,
-    );
-
-    // Resolve the requirements.
-    let resolution = pip::operations::resolve(
-        requirements,
-        constraints,
-        overrides,
-        dev,
-        source_trees,
-        None,
-        &extras,
-        preferences,
-        EmptyInstalledPackages,
-        &hasher,
-        &Reinstall::default(),
-        upgrade,
-        None,
-        None,
-        python_requirement,
-        &client,
-        &flat_index,
-        &index,
-        &build_dispatch,
-        concurrency,
-        options,
-        printer,
-        preview,
-    )
-    .await?;
+    // Try resolving from the lockfile.
+    let resolution = match lock {
+        Some(lock) => {
+            // Prefill the index with the lockfile.
+            let index = lock.to_index(workspace)?;
+
+            // Create a build dispatch.
+            //
+            // When resolving from the lockfile we can still download/install new distributions,
+            // but we rely on the lockfile for the metadata of any existing distributions. If we
+            // have any outdated metadata we fall back to a clean resolve.
+            let build_dispatch = BuildDispatch::new(
+                &client,
+                cache,
+                interpreter,
+                index_locations,
+                &flat_index,
+                &index,
+                &git,
+                &in_flight,
+                index_strategy,
+                setup_py,
+                config_setting,
+                build_isolation,
+                link_mode,
+                build_options,
+                exclude_newer,
+                concurrency,
+                preview,
+            );
+
+            // TODO(ibraheem): avoid cloning the manifest here
+            pip::operations::resolve(
+                requirements.clone(),
+                constraints.clone(),
+                overrides.clone(),
+                dev.clone(),
+                source_trees.clone(),
+                None,
+                &extras,
+                preferences.clone(),
+                EmptyInstalledPackages,
+                &hasher,
+                &Reinstall::default(),
+                upgrade,
+                None,
+                None,
+                python_requirement.clone(),
+                &client,
+                &flat_index,
+                &index,
+                &build_dispatch,
+                concurrency,
+                options,
+                printer,
```

Comment:
Should stderr be suppressed here? Thinking there might be duplicate warnings that might be printed if resolution from a lockfile fails.

Reply:
Yeah, we probably want to suppress everything here so that if resolution succeeds, we don't log anything about it.

Reply:
I'm not sure we should supress everything because this resolve might still involve downloading/building new distributions?

Reply:
When would we need to download / build a new distribution? The only case in which we build a distribution during resolve is to create a wheel from a source dist, but don't we already have the metadata?

Reply:
We're still doing a regular resolve with the default provider, only existing immutable metadata is prefilled. New dependencies can still be fetched and built.

Reply:
Yeah not sure how to solve this. I think I'd need to see the output in various cases to develop an opinion on it...

Reply:
I guess the only output is like... anything related to building (which we want to keep, right?)... then the summary message (which means we succeeded, so that's fine)... then warnings? But we only print warnings (1) if the resolution fails, as hints to the user (so those wouldn't be shown), and then (2) via "diagnostics" on the resolution (look for `operations::diagnose_resolution` calls). So actually this is probably totally fine as-is?